### PR TITLE
Add structured dataset type checker

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
Add structured dataset type checker

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flytekit/pull/785

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/compiler/validators/typing.go [hunk 1]
```diff
@@ -130,6 +134,46 @@ func (t schemaTypeChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
...
+		nameToTypeMap[column.Name] = column.LiteralType
+	}
+
+	// Check that the downstream structured dataset is a strict sub-set of the upstream structured dataset.
+	for _, column := range t.literalType.GetStructuredDatasetType().Columns {
+		upstreamType, ok := nameToTypeMap[column.Name]
+		if !ok {
+			return false
+		}
+		if !getTypeChecker(upstreamType).CastsFrom(column.LiteralType) {
```
Comment (EngHabu) [inline pkg/compiler/validators/typing.go:170]: Should this be the inverse? Get type checker of column.LiteralType

Comment (pingsutw) [reply pkg/compiler/validators/typing.go]: yeah, you're right. updated it.

File: Dockerfile [hunk 2]
```diff
@@ -18,7 +18,7 @@ COPY . /go/src/github.com/flyteorg/flytepropeller/
 
 # This 'linux_compile' target should compile binaries to the /artifacts directory
 # The main entrypoint should be compiled to /artifacts/flytepropeller
-RUN make linux_compile
+RUN go mod vendor && make linux_compile
```
Comment (wild-endeavor) [inline Dockerfile:21]: go mod download maybe?  we don't vendor anymore.  do you remember what the error was?  you shouldn't have to do this.

File: pkg/compiler/validators/typing.go [hunk 3]
```diff
@@ -130,6 +134,46 @@ func (t schemaTypeChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
...
+// cases.
+//
+//    1. The downstream structured dataset has no column types specified.  In such a case, it accepts all structured dataset input since it is
+//       generic.
+//
+//    2. The downstream structured dataset has a subset of the upstream columns and they match perfectly.
+//
+func (t structuredDatasetChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
+	// structured dataset are nullable
+	if isVoid(upstreamType) {
```
Comment (wild-endeavor) [inline pkg/compiler/validators/typing.go:147]: is this common?

File: pkg/compiler/validators/typing.go [hunk 4]
```diff
@@ -130,6 +134,46 @@ func (t schemaTypeChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
...
+		return true
+	}
+	structuredDatasetType := upstreamType.GetStructuredDatasetType()
```
Comment (wild-endeavor) [inline pkg/compiler/validators/typing.go:150]: I think we want to allow for FlyteSchema as well actually.

**Review by wild-endeavor — COMMENTED** [review]
could you get @EngHabu to take a look?  i'm not owner on this repo anyways

File: Dockerfile (hunk 2 above)
Comment (pingsutw) [reply Dockerfile:21]: This error only happens when I replace the idl version with another branch. like 
`replace github.com/flyteorg/flyteidl => github.com/flyteorg/flyteidl v0.21.13-0.20211206210958-804fa1685264
`
will remove this before we merge the PR

File: pkg/compiler/validators/typing.go (hunk 3 above)
Comment (pingsutw) [reply pkg/compiler/validators/typing.go:147]: we did the same thing in schema, so I added this

File: Dockerfile (hunk 2 above)
Comment (pingsutw) [reply Dockerfile:21]: Error message:
```bash
#15 0.319 	google.golang.org/grpc@v1.36.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#15 0.319 	google.golang.org/protobuf@v1.25.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#15 0.319 	k8s.io/api@v0.20.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#15 0.319 	k8s.io/apimachinery@v0.20.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#15 0.319 	k8s.io/client-go@v0.20.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#15 0.319 	k8s.io/klog@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#15 0.319 	sigs.k8s.io/controller-runtime@v0.8.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
#15 0.319 	github.com/aws/amazon-sagemaker-operator-for-k8s: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
#15 0.319 	github.com/flyteorg/flyteidl: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
#15 0.319 	github.com/flyteorg/flyteplugins: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
#15 0.319 
#15 0.319 	To ignore the vendor directory, use -mod=readonly or -mod=mod.
#15 0.319 	To sync the vendor directory, run:
#15 0.319 		go mod vendor
#15 0.320 make: *** [Makefile:14: linux_compile] Error 1
------
executor failed running [/bin/sh -c make linux_compile]: exit code: 2
make: *** [docker_build] Error 1
```

File: pkg/compiler/validators/typing.go (hunk 4 above)
Comment (pingsutw) [reply pkg/compiler/validators/typing.go:152]: done

Comment (pingsutw) [reply pkg/compiler/validators/typing.go:152]: I've tested it. check this [example](https://gist.github.com/pingsutw/20bba3ead52a9881ddb02b983efe1a5d)

image

**Review by wild-endeavor — COMMENTED** [review]
wait sorry - can we add a check for format also?  format should be `==`.  but case-insensitive.

**Comment by EngHabu** [comment]
I see a lot of changes in the PR that are not related... you may have to start a new branch if this is borked

**Comment by pingsutw** [comment]
sorry, will remove them

**Pull request closed without merging**